packages/remix-dev/config/format.ts

```diff
@@ -90,9 +70,95 @@ export function formatRoutesAsJsx(routeManifest: RouteManifest) {
     return routes.length > 0;
   }
 
-  handleRoutesRecursive();
+  handleRoutesRecursive(routes);
 
   output += "\n</Routes>";
 
   return output;
 }
+
+interface BaseManifestRoute {
+  id: string;
+  path?: string;
+  parentId?: string;
+}
+
+interface BaseHierarchyRoute {
+  id: string;
+  path?: string;
+  children?: BaseHierarchyRoute[];
+}
+
+/**
+ * NOTE: This function is duplicated in remix-dev, remix-react, and
+ * remix-server-runtime so if you make changes please make them in all 3
+ * locations. We'll look into DRY-ing this up after we layer Remix on top of
+ * react-router@6.4
+ *
+ * Generic reusable function to convert a manifest into a react-router style
+ * route hierarchy.  For use in server-side and client-side route creation,
+ * as well and `remix routes` to keep them all in sync.
+ *
+ * This also handles inserting "folder" parent routes to help disambiguate
+ * between pathless layout routes and index routes at the same level
+ *
+ * @param manifest     Map of string -> Route Object
+ * @param createRoute  Function to create a hierarchical route given a manifest
+ *                     ignoring children
+ * @returns
+ */
+function createHierarchicalRoutes<
+  ManifestRoute extends BaseManifestRoute,
+  HierarchyRoute extends BaseHierarchyRoute
+>(
+  manifest: Record<string, ManifestRoute>,
+  createRoute: (r: ManifestRoute) => HierarchyRoute
+) {
```

Comment:
Generic function to convert a flat manifest to a hierarchical structure, accepting a function to convert the manifest route into the hierarchical route.  Ideally this would be shared amongst remix-dev/remix-server-runtime/remix-react but is duped for now.  This contains the logic to insert parent routes as needed to disambiguate pathless routes